src/lib/Controller/AutoDetect.cpp

```diff
@@ -0,0 +1,109 @@
+#include "targets.h"
+
+#if defined(PLATFORM_ESP32) && defined(TARGET_TX)
+
+#include "AutoDetect.h"
+#include "CRSFController.h"
+#include "PPMController.h"
+#include "devController.h"
+#include "logging.h"
+
+#include <driver/rmt.h>
+
+const rmt_channel_t PPM_RMT_CHANNEL = RMT_CHANNEL_0;
+const auto RMT_TICKS_PER_US = 10;
+
+void AutoDetect::Begin()
+{
+    auto divisor = 80 / RMT_TICKS_PER_US;
+
+    rmt_config_t rmt_rx_config = RMT_DEFAULT_CONFIG_RX(static_cast<gpio_num_t>(GPIO_PIN_RCSIGNAL_RX), PPM_RMT_CHANNEL);
+    rmt_rx_config.clk_div = divisor;
+    rmt_rx_config.rx_config.filter_ticks_thresh = 1;
+    rmt_rx_config.rx_config.idle_threshold = 100;
+    rmt_config(&rmt_rx_config);
+    rmt_driver_install(PPM_RMT_CHANNEL, 1000, 0);
+
+    rmt_get_ringbuf_handle(PPM_RMT_CHANNEL, &rb);
+    rmt_rx_start(PPM_RMT_CHANNEL, true);
+    input_detect = 0;
+}
+
+void AutoDetect::End()
+{
+    rmt_driver_uninstall(PPM_RMT_CHANNEL);
+}
+
+bool AutoDetect::IsArmed()
+{
+    return false;
+}
+
+void AutoDetect::startPPM()
+{
+    Controller *ppm = new PPMController();
+    ppm->setRCDataCallback(RCdataCallback);
+    ppm->registerParameterUpdateCallback(RecvParameterUpdate);
+    ppm->registerCallbacks(connected, disconnected, RecvModelUpdate);
+    ppm->Begin();
+    ppm->setPacketInterval(RequestedRCpacketInterval);
+    delete this;
+    controller = ppm;
+}
+
+void AutoDetect::startCRSF()
+{
+    Controller *crsf = new CRSFController();
+    crsf->setRCDataCallback(RCdataCallback);
+    crsf->registerParameterUpdateCallback(RecvParameterUpdate);
+    crsf->registerCallbacks(connected, disconnected, RecvModelUpdate);
+    crsf->Begin();
+    crsf->setPacketInterval(RequestedRCpacketInterval);
+    delete this;
+    controller = crsf;
```

Comment:
The code between PPM and CRSF is identical apart from the `new`, refactor to use a common function to do all the stuff?
```cpp
void AutoDetect::migrateTo(Controller *that)
{
    that->setRCDataCallback(RCdataCallback);
    that->registerParameterUpdateCallback(RecvParameterUpdate);
    that->registerCallbacks(connected, disconnected, RecvModelUpdate);
    that->Begin();
    that->setPacketInterval(RequestedRCpacketInterval);
    delete this;
    controller = that;
}

void AutoDetect::startCRSF()
{
    migrateTo(new CRSFController()):
}
```

Reply:
Done